src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs

```diff
@@ -81,7 +76,7 @@ public partial class RegexGenerator
             int? matchTimeout = null;
             foreach (AttributeData attributeData in boundAttributes)
             {
-                if (attributeData.AttributeClass?.Equals(regexGeneratorAttributeSymbol) != true)
+                if (attributeData.AttributeClass?.ToDisplayString()?.Equals(RegexGeneratorAttributeName, StringComparison.Ordinal) != true)
```

Comment:
I am wondering whether this change is actually equivalent. Previously we were taking a symbol object from the compilation, and checking for equality. Now we are comparing the symbols by name. The only case I imagine it would cause problems is if there was another `System.Text.RegularExpressions.RegexGeneratorAttribute` in user code, but we are already checking tha attribute's validity later.

Or maybe checking the name is the correct way, given that we were using `GetBestTypeByMetadataName`, an internal method copy-pasted from Roslyn's sources.

Reply:
> My question is whether this proposed logic is far too flexible and accepts things it shouldn't.

It seems like it to me, and also potentially more expensive if it has to materialize strings for all of the attributes it sees. I don't know if those are cached, but I do know I was previously asked not to do this in roslyn-analyzers.

@sharwell and @chsienki should provide the authoritative answer, though.

Reply:
It should be fine if you check `AttributeClass?.Name` first.

Reply:
So this is an interesting question. It does have a slightly different semantics and the short answer is just use `SemanticModel.Compilation` and you can keep the same behavior as before. The longer answer comes down to how conservative you want the generator to be with which attributes you accept. 

`GetBestTypeByMetadataName` has a defined search order when there are ambiguities (source first, then references considering visibility), so you know *which* attribute it will select in the case of multiples, but it still doesn't guarentee that you get the one you're expecting: as a user I can define a new` System.Text.RegularExpressions.RegexGeneratorAttribute` in source, and the current implementation will pick it here. If I also use that attribute on my code the checks will succeed, but it may not have all the information required on, leading to a `SYSLIB1040` later on.

Using `GetTypeByMetadataName` will only return when there is *exactly one* matching type, so its better for the case of getting what you expect (you can still muck around with ref assemblies and break it, but its much more deliberate).

Using the name check (from the semantic model, after alias resolution) is the most permissive: basically anything that matches on the metadata name will be considered. Given that you actually check the attribute for well-formed-ness later on, I think this is the most appropriate approach. Its a bit like the (Robustness principle)[https://en.wikipedia.org/wiki/Robustness_principle], be liberal in the expected input from the compilation, and do the best generation you can. 

In terms of the perf, it's proabably a wash, but wouldn't want to say one way or the other without actually profiling it first.
`GetBestTypeByMetadataName` has a defined search order when there are ambiguities (source first, then references considering visibility), so you know *which* attribute it will select in the case of multiples, but it still doesn't guarentee that you get the one you're expecting: as a user I can define a new `System.Text.RegularExpressions.RegexGeneratorAttribute` in source, and the current implementation will pick it here. If I also use that attribute on my code the checks will succeed, but it may not have all the information selected later on, leading to a SYSLIB1040 later on.

Using `GetTypeByMetadataName` will only return when there is *exactly one* matching type, so its better for the case of getting what you expect (you can still muck around with ref assemblies and break it, but its much more deliberate).

Using name only is the most permissive: basically anything that matches on the metadata name will be considered. Given that you actually check the attribute for well-formed-ness later on, I think this is the most appropriate approach. Its a bit like the [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle), be liberal in the expected input from the compilation, and do the best generation you can. 

In terms of the perf, it's proabably a wash, but wouldn't want to say one way or the other without actually profiling it first.

Reply:
@chsienki It looks like you posted a bunch of content twice in copy/paste fashion, but it's not exactly identical. I added a separator between the two sections with overlapping content in case you want to edit it.

Reply:
@sharwell Weird, it wasn't showing that for me. Thanks. edited.

Reply:
Thanks @chsienki, I had stumbled upon `SemanticModel.Compilation` but thought using it was a cheat because it would break incrementality. Since it's OK in this case I went back to the old way of matching symbols.